Assets/MixedRealityToolkit.SDK/Experimental/Features/Support/Scripts/BaseServiceManager.cs

```diff
@@ -0,0 +1,395 @@
+
+                IMixedRealityDataProvider[] providers = dataProviders.ToArray();
+                for (int i = 0; i < providers.Length; i++)
+                {
+                    providers[i]?.LateUpdate();
+                }
+            }
+        }
+
+
+        protected virtual void OnEnable()
+        {
+            if (Application.isPlaying)
+            {
+                IReadOnlyList<IMixedRealityService> serviceSnapshot = new List<IMixedRealityService>(registeredServices.Values);
+                for (int i = 0; i < serviceSnapshot.Count; i++)
+                {
+                    serviceSnapshot[i]?.Enable();
+                }
+
+                IMixedRealityDataProvider[] providers = dataProviders.ToArray();
+                for (int i = 0; i < providers.Length; i++)
+                {
+                    providers[i]?.Enable();
+                }
+            }
+        }
+
+        protected virtual void OnDisable()
+        {
+            if (Application.isPlaying)
+            {
+                IMixedRealityDataProvider[] providers = dataProviders.ToArray();
+                for (int i = 0; i < providers.Length; i++)
+                {
+                    providers[i]?.Disable();
+                }
+
+                IReadOnlyList<IMixedRealityService> serviceSnapshot = new List<IMixedRealityService>(registeredServices.Values);
+                for (int i = 0; i < serviceSnapshot.Count; i++)
+                {
+                    serviceSnapshot[i]?.Disable();
+                }
+            }
+        }
+
+        protected virtual void OnDestroy()
+        {
+            IMixedRealityDataProvider[] providers = dataProviders.ToArray();
+            for (int i = 0; i < providers.Length; i++)
+            {
+                providers[i]?.Disable(); // Disable before destroy to ensure the data provider has time to get in a good state.
+                providers[i]?.Destroy();
+            }
+            // Clear the actual collection
+            dataProviders.Clear();
+
+            IReadOnlyList<IMixedRealityService> serviceSnapshot = new List<IMixedRealityService>(registeredServices.Values);
+            for (int i = 0; i < serviceSnapshot.Count; i++)
+            {
+                serviceSnapshot[i].Disable(); // Disable before destroy to ensure the service has time to get in a good state.
+                serviceSnapshot[i].Destroy();
+            }
+            // Clear the actual collection
+            registeredServices.Clear();
+        }
+
+        #endregion MonoBehaviour implementation
+
+        #region IMixedRealityServiceRegistrar implementation
+
+        /// <inheritdoc />
+        public T GetDataProvider<T>(string name = null) where T : IMixedRealityDataProvider
+        {
+            Type interfaceType = typeof(T);
+            T provider = default(T);
+
+            IMixedRealityDataProvider[] providers = dataProviders.ToArray();
+            for (int i = 0; i < providers.Length; i++)
```

Comment:
Just iterate over the list directly instead of allocating another array for this purpose (same comment everywhere else)

Reply:
Otherwise this needlessly allocates memory for real value (you're not modifying the provider list here, which I guess would be possible elsewhere)

---
In reply to: [301757810](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5233#discussion_r301757810) [](ancestors = 301757810)

Reply:
fixed